msamp/nn/functional.py

```diff
@@ -0,0 +1,178 @@
+        weight_fp8 = weight.cast(Dtypes.kfloat8_e4m3)
+
+        ctx.input_fp8 = input_fp8
+        ctx.input_fp8.requires_grad = input.requires_grad
+        ctx.weight_fp8 = weight_fp8
+        ctx.weight = weight
+
+        output_dtype = dtype_holder.dtype
+        output_qtype = Dtypes.dtype_to_qtype[output_dtype]
+
+        ctx.output_dtype = output_dtype
+        ctx.output_qtype = output_qtype
+
+        out = Gemm.fp8_gemm(weight_fp8, input_fp8, output_qtype, use_split_accumulator=False)
+        return out
+
+    @staticmethod
+    def backward(ctx, output_grad):
+        """Backward function.
+
+        Args:
+            ctx: Context to get the data stored in forward pass.
+            output_grad (torch.Tensor): Output gradient tensor.
+
+        Returns:
+            tuple (torch.Tensor, torch.Tensor, torch.Tensor, torch.Tensor): The gradients of the arguments
+                in forward function. None if no gradient.
+        """
+        # pytorch has a bug that output_grad.strides is 0. Use .contiguous() to fix it.
+        output_grad = output_grad.contiguous()
+
+        # We assign gradients to x.grad directly.
+        metas = ctx.metas
+        ograd_meta = metas['ograd']
+        wgrad_meta = metas['wgrad']
+        ograd_fp8 = output_grad.cast(Dtypes.kfloat8_e5m2, meta=ograd_meta)
+
+        if ctx.input_fp8.requires_grad:
+            weight_fp8_t = ctx.weight_fp8.t().contiguous()
+            input_grad = Gemm.fp8_gemm(weight_fp8_t, ograd_fp8, ctx.output_qtype, use_split_accumulator=True)
+        else:
+            input_grad = None
+
+        if ctx.weight.requires_grad:
+            ograd_fp8_t = ograd_fp8.t().contiguous()
+            input_fp8_t = ctx.input_fp8.t().contiguous()
+            wgrad_qtype = ctx.output_qtype
+            # compute weight gradient
+            if ctx.weight.grad is None:
+                wgrad = Gemm.fp8_gemm(
+                    input_fp8_t,
+                    ograd_fp8_t,
+                    wgrad_qtype,
+                    use_split_accumulator=True,
+                )
+            else:
+                # gradient accumulation, old_wgrad is FP32 or FP16 without tensor scaling.
+                old_wgrad = ctx.weight.grad.to(ctx.output_dtype)
+                wgrad = Gemm.fp8_gemm(
+                    input_fp8_t,
+                    ograd_fp8_t,
+                    wgrad_qtype,
+                    accumulate=True,
+                    out=old_wgrad,
+                    use_split_accumulator=True,
+                )
+                del old_wgrad
+
+            # wgrad above this line is torch.Tensor w/o tensor scaling
+            wgrad = wgrad.cast(Dtypes.kfloat8_e4m3, meta=wgrad_meta, sync=True)
+            if DistUtil.get_world_size() > 1:
+                model_state.ready_to_all_reduce_grads = True
+
+            ctx.weight.backward_grad_update(wgrad)
+
+        return input_grad, None, None, None
+
+
+class FunctionalOverider:
```

Comment:
Can we move "the overrider of F.Linear" to TorchOverrider?

Reply:
No. `TorchOverrider` is in `common/tensor/tensor.py`. It is unsuitable to import `msamp.nn` in `common.tensor`.